### PR TITLE
Distinguish more between bounds correct vs bounds loaded

### PR DESCRIPTION
**Pull request by abadams**
Currently computation bounds inference just respects box_touched, which means that for something like `select(x < 100, f(x), g(x))`, f and g are computed over the entire domain. It would be better if we allocated f and g over the entire domain, but just computed f before 100, and just computed g after 100. 

This PR achieves this by treating selects as if_then_else instead when doing computation bounds inference. In allocation bounds inference they're still selects. Also fixed a bug in Bounds along the way (it was checking that the solver terminated with a single copy of the variable, but not that the thing on the LHS was just the variable, and not, e.g. `x%2`).

I'm a little nervous about compile time implications. I would appreciate if people with any particularly sensitive pipelines could try this out.

### CODE REVIEW

File: src/Bounds.cpp [hunk 1]
```diff
@@ -2061,17 +2061,21 @@ class BoxesTouched : public IRGraphVisitor {
...
                     const GE *ge = solved.as<GE>();
                     const EQ *eq = solved.as<EQ>();
-                    Expr rhs;
-                    if (lt) {rhs = lt->b;}
-                    if (le) {rhs = le->b;}
-                    if (gt) {rhs = gt->b;}
-                    if (ge) {rhs = ge->b;}
-                    if (eq) {rhs = eq->b;}
+                    Expr lhs, rhs;
+                    if (lt) {lhs = lt->a; rhs = lt->b;}
```
Comment (steven-johnson) [inline src/Bounds.cpp]: super-nit: would it be worthwhile to make these if-else since they are all mutually exclusive?

**Comment by abadams** [comment]
It's not urgent. I need this for another feature I'm working on, but I can just merge this into my other branch.

**Comment by steven-johnson** [comment]
Let me pull current master into our code (so that CSE, etc fixes are already in place), then I'll do a test integrate of this if you think it's ready

**Comment by steven-johnson** [comment]
Pulling this into google, I get many hundreds of failures, but all seem to be variants of

```
The pure definition  of Function foo calls function bar in an unbounded way in dimension 0
```

or

```
Symbol not found: baz$3.s0.fx.max
```

**Comment by abadams** [comment]
I'm at a loss for what might be failing. Are any of the failures small enough to turn into a repro? Otherwise I could also cut this into several smaller PRs and we can bisect what's causing the problem.

**Comment by steven-johnson** [comment]
No idea yet -- I was doing bisection on my end but have been sidetracked by more important matters. I hope to get back to it soon.

**Comment by steven-johnson** [comment]
Still getting oodles of 'unbounded' where I didn't before, and I have high-priority tasks ahead of this for the next couple of days. If it's easy to split this PR into smaller pieces that might be an easier test.

**Comment by steven-johnson** [comment]
FYI: if I disable the changes in BoundInference.cpp (but leave the rest), my mystery failures go away.

**Comment by abadams** [comment]
I split the changes in BoundsInference.cpp into two and opened #3859

**Comment by abadams** [comment]
Updated to a form that passes the repro from the other PR (#3859)

**Comment by steven-johnson** [comment]
To test this in Google, I probably need to land #3867 first, then pull Halide in, to avoid unrelated failures.

**Comment by steven-johnson** [comment]
Running a test of this inside Google, and so far so good. Will do a more thorough test and hopefully we can land this later today.

**Comment by abadams** [comment]
Excellent. Getting zero-copy reshapes and concats from the store_with PR would be a good feature to be able to merge, but the concats rely on this PR to work out cleanly.

**Comment by steven-johnson** [comment]
Not seeing any new failures due to this. OK to land.

**Pull request merged**